bin/manage.sh

```diff
@@ -1,67 +1,130 @@
 #!/bin/bash
 
 MASTER=null
+CONSUL_HOST=${CONSUL}
+CONSUL_AGENT=${CONSUL_AGENT:=false}
 
-if [[ -z ${CONSUL} ]]; then
+readonly lockPath=service/elasticsearch-master/locks/master
+
+if [ $CONSUL_AGENT != false ]; then
+    CONSUL_HOST='localhost'
+fi
+
+if [[ -z $CONSUL_HOST ]]; then
     echo "Missing CONSUL environment variable"
     exit 1
 fi
 
+consulCommand() {
+    consul-cli --quiet --consul="${CONSUL_HOST}:8500" $*
+}
+
 preStart() {
-    # happy path is that there's a master available and we can cluster
-    configureMaster
-
-    # data-only nodes can only loop until there's a master available
-    if [ ${ES_NODE_MASTER} == false ]; then
-        while true
-        do
-            sleep 1.7
-            configureMaster
+    logDebug "preStart"
+}
+
+onStart() {
+    logDebug "onStart"
+
+    waitForLeader
+
+    getRegisteredServiceName
+    if [[ "${registeredServiceName}" == "elasticsearch-data" ]]; then
+
+        # wait for a healthy master
+        local i
+        for (( i = 0; i < ${MASTER_WAIT_TIMEOUT-60}; i++ )); do
+            getServiceAddresses "elasticsearch-master"
+            if [[ ${serviceAddresses} ]]; then
+                MASTER=$serviceAddresses
+                break
+            fi
+            sleep 1
         done
-        exit 0
-    fi
 
-    # for a master+data node, we'll retry to see if there's another
-    # master in the cluster in the process of starting up. But we
-    # bail out if we exceed the retries and just bootstrap the cluster
-    if [ ${ES_NODE_DATA} == true ]; then
-        local n=0
-        until [ $n -ge 2 ]
-        do
-            sleep 1.7
-            configureMaster
-            n=$((n+1))
+    else
+
+        # wait for a healthy master
```

Comment:
> It is ugly how it just keeps hitting consul every second to see if there are any other services registered as 'elasticsearch-master'. Maybe creating a lock is a good idea?

You mean here? That's only during startup, so while I suppose it's ugly it's also rather simple and it goes away after the initial startup.  How does creating a lock help us out here?